errorcodes/errorcodes.go

```diff
@@ -0,0 +1,280 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package errorcodes
+
+import (
+	"errors"
+	"fmt"
+
+	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/runtime/protoiface"
+)
+
+type Reason string
+
+const (
+	ErrorCodesFeatureMetadataKey = "error_codes_feature"
+	Owner                        = "components-contrib"
+	Domain                       = "dapr.io"
+)
+
+var (
+	NoReason                       = Reason("NO_REASON_SPECIFIED")
+	StateETagMismatchReason        = Reason("DAPR_STATE_ETAG_MISMATCH")
+	StateETagInvalidReason         = Reason("DAPR_STATE_ETAG_INVALID")
+	TopicNotFoundReason            = Reason("DAPR_TOPIC_NOT_FOUND")
+	SecretKeyNotFoundReason        = Reason("DAPR_SECRET_KEY_NOT_FOUND")
+	ConfigurationKeyNotFoundReason = Reason("DAPR_CONFIG_KEY_NOT_FOUND")
+)
+
+type ResourceInfo struct {
+	ResourceType string
+	ResourceName string
+}
+
+// call this function to apply option
+type ErrorOption func(*DaprError)
+
+type DaprError struct {
+	err          error
+	description  string
+	reason       Reason
+	resourceInfo *ResourceInfo
+	metadata     map[string]string
+}
+
+// EnableComponentErrorCode stores an indicator for
+// components to determine if the ErrorCodes Feature
+// is enable.
+func EnableComponentErrorCode(metadata map[string]string) {
+	if metadata != nil {
+		metadata[ErrorCodesFeatureMetadataKey] = "true"
+	}
+}
+
+func featureEnabled(metadata map[string]string) bool {
```

Comment:
If metadata can be nil above, it can be nil here too.

Reply:
Something I learned from one of the comments from other PRs about nil maps in Go is that lookup operations can be performed safely and will return the zero value of the entry value type. In this case false. Only adding an entry to a nil map is not allowed.